contracts/ERC20MultiDelegate.sol

```diff
@@ -0,0 +1,434 @@
+            "DepositMulti: You should pick at least one delegatee"
+        );
+
+        uint256[] memory ids = new uint256[](delegateesLength);
+        uint256[] memory amounts = new uint256[](delegateesLength);
+
+        for (uint256 index = 0; index < delegateesLength; index++) {
+            address delegatee = delegateeAmounts[index].delegatee;
+            uint256 amount = delegateeAmounts[index].amount;
+
+            createProxyDelegatorAndTransfer(delegatee, amount);
+
+            ids[index] = uint256(uint160(delegatee));
+            amounts[index] = amount;
+        }
+
+        mintBatch(msg.sender, ids, amounts);
+    }
+
+    /**
+     * @dev Re-delegates voting power between delegatees.
+     * @param sources The list of source delegatee addresses and amounts to withdraw.
+     * @param targets The list of target delegatee addresses and amounts to redeposit.
+     * The remaning part if any, will be withdrawn to the user account
+     */
+    function reDeposit(
+        DelegateeAmount[] calldata sources,
+        DelegateeAmount[] calldata targets
+    ) external {
+        uint256 sourcesLength = sources.length;
+        uint256 targetsLength = targets.length;
+
+        require(
+            sourcesLength > 0 && targetsLength > 0,
+            "ReDeposit: You should pick at least one source and one target delegatee"
+        );
+
+        ReDepositRecord memory record = ReDepositRecord({
+            sourceIds: new uint256[](sourcesLength),
+            targetIds: new uint256[](targetsLength),
+            withdrawnAmounts: new uint256[](sourcesLength),
+            redepositedAmounts: new uint256[](targetsLength),
+            sourceIndex: 0,
+            targetIndex: 0,
+            remainingSourceAmount: sources[0].amount,
+            remainingTargetAmount: targets[0].amount,
+            totalRedepositedAmount: 0
+        });
+
+        _reDeposit(sources, targets, record);
+
+        require(
+            record.remainingTargetAmount == 0,
+            "ReDeposit: Total target amount cannot be greater than source amount"
+        );
+
+        // Reimburses any remaining source amounts back to the delegator after the re-deposit process.
+        if (record.remainingSourceAmount > 0) {
+            _reimburse(record, sources, sourcesLength, msg.sender);
+        }
+
+        burnBatch(msg.sender, record.sourceIds, record.withdrawnAmounts);
+        mintBatch(msg.sender, record.targetIds, record.redepositedAmounts);
+    }
+
+    /**
+     * @dev Executes the re-deposit process for multiple source and target delegatees.
+     * @param sources The list of source delegatees with the amounts to withdraw.
+     * @param targets The list of target delegatees with the amounts to re-deposit.
+     * @param record The current state of the re-deposit process.
+     */
+    function _reDeposit(
+        DelegateeAmount[] calldata sources,
+        DelegateeAmount[] calldata targets,
+        ReDepositRecord memory record
```

Comment:
I think you should integrate the outer loop into this function, so the internal state doesn't need to be preserved in a struct.